ansible/roles/data-products-deploy/templates/submit-script.j2

```diff
@@ -0,0 +1,88 @@
+        -s|--parallelisation)   shift 
+                                parallelisation=$1            
+        ;;
+        -sd|--startDate)   shift 
+                            start_date=$1            
+        ;;
+        -ed|--endDate)     shift 
+                            end_date=$1            
+        ;;
+        -h|--sparkMaster)   shift 
+                            sparkMaster=$1            
+        ;;
+        *) break
+    esac
+    shift
+done
+
+if [ -z "$sparkMaster" ]; then sparkMaster="local[*]"; else sparkMaster="$sparkMaster"; fi
+
+if [ "$type" = "via-partition" ]; then
+    # get partitions from kafka metadata
+    if [ -z "$partitions" ]; then partitions=`$KAFKA_HOME/bin/kafka-topics.sh --zookeeper $zookeeper --describe --topic $unique_topic   | awk 'NR==1{print $2}' |  awk -F":" '{print $2}'`; else partitions=$partitions; fi
+    endPartitions=`expr $partitions - 1`
+    if [ -z "$parallelisation" ]; then parallelisation=1; else parallelisation=$parallelisation; fi
+    # add partitions to config and start jobs
+    for i in $(seq 0 $parallelisation $endPartitions)
+        do 
+            # add partitions to config
+            partitionString="\"delta\":0,\"partitions\":[$(seq -s , $i `expr $i + $parallelisation - 1`)]"
+            if [ -z "$start_date" ]; then
+                job_config=$(config $job)
+                finalConfig=${job_config/'"delta":0'/$partitionString}
+                echo $finalConfig
+                echo "Running $job by partitions."
+                nohup $SPARK_HOME/bin/spark-submit --master $sparkMaster --jars $(echo ${libs_path}/lib/*.jar | tr ' ' ','),$MODELS_HOME/analytics-framework-2.0.jar,$MODELS_HOME/scruid_2.11-2.3.2.jar,$MODELS_HOME/batch-models-2.0.jar --class org.ekstep.analytics.job.JobExecutor $MODELS_HOME/batch-models-2.0.jar --model "$job" --config "$finalConfig" >> "$DP_LOGS/$today-job-execution.log" 2>&1
+            else 
+                job_config=$(config $job '__endDate__')
+                finalConfig=${job_config/'"delta":0'/$partitionString}
+                echo $finalConfig
+                echo "Running $job by partitions via Replay-Supervisor."
+                nohup $SPARK_HOME/bin/spark-submit --master $sparkMaster --jars $MODELS_HOME/analytics-framework-2.0.jar,$MODELS_HOME/scruid_2.11-2.3.2.jar --class org.ekstep.analytics.job.ReplaySupervisor $MODELS_HOME/batch-models-2.0.jar --model "$job" --fromDate "$start_date" --toDate "$end_date" --config "$finalConfig" >> "$DP_LOGS/$end_date-$job-replay.log" 2>&1
+            fi
+        done
+elif [ "$type" = "job-manager" ]; then
+    # push job-config to kafka      
+    echo "Submitting $job to job-manager."
+    echo '{ "model" :' \"$job\" ',' ' "config": ' "$job_config" '}' > /tmp/job-request.json
+    cat /tmp/job-request.json | $KAFKA_HOME/bin/kafka-console-producer.sh --broker-list $brokerList --topic $job_topic >> "$DP_LOGS/$today-job-execution.log" 2>&1
+else
+    if [ -z "$start_date" ]; then
```

Comment:
The start_date and end_date needs to be computed outside and the spark-submit called only once.

Reply:
refactored to use spark-submit only once